crates/llm-remote-http/src/lib.rs

```diff
@@ -0,0 +1,186 @@
+    generated_token_count: u32,
+}
+#[derive(Deserialize)]
+struct InferResponseBody {
+    text: String,
+    usage: InferUsage,
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all(deserialize = "camelCase"))]
+struct EmbeddingUsage {
+    prompt_token_count: u32,
+}
+
+#[derive(Deserialize)]
+struct EmbeddingResponseBody {
+    embeddings: Vec<Vec<f32>>,
+    usage: EmbeddingUsage,
+}
+
+#[async_trait]
+impl LlmEngine for RemoteHttpLlmEngine {
+    fn set_allowed_models(&mut self, models: HashSet<String>) {
+        self.allowed_models = models;
+    }
+
+    async fn infer(
+        &mut self,
+        model: wasi_llm::InferencingModel,
+        prompt: String,
+        params: InferencingParams,
+    ) -> Result<wasi_llm::InferencingResult, wasi_llm::Error> {
+        if !self.allowed_models.contains(&model) {
+            return Err(access_denied_error(&model));
+        }
+        let client = self.client.get_or_insert_with(Default::default);
+
+        let mut headers = HeaderMap::new();
+        headers.insert(
+            "authorization",
+            HeaderValue::from_str(&format!("bearer {}", self.auth_token)).map_err(|_| {
+                wasi_llm::Error::RuntimeError("Failed to create authorization header".to_string())
+            })?,
+        );
+        let inference_options = InferRequestBodyParams {
+            max_tokens: params.max_tokens,
+            repeat_penalty: params.repeat_penalty,
+            repeat_penalty_last_n_token_count: params.repeat_penalty_last_n_token_count,
+            temperature: params.temperature,
+            top_k: params.top_k,
+            top_p: params.top_p,
+        };
+        let body = serde_json::to_string(&json!({
+            "model": model,
+            "prompt": prompt,
+            "options": inference_options
+        }))
+        .map_err(|_| wasi_llm::Error::RuntimeError("Failed to serialize JSON".to_string()))?;
+
+        let resp = client
+            .request(
+                http::Method::POST,
+                self.url.join("/infer").map_err(|_| {
+                    wasi_llm::Error::RuntimeError("Failed to create URL".to_string())
+                })?,
+            )
+            .headers(headers)
+            .body(body)
+            .send()
+            .await;
+
+        match resp {
+            Ok(resp) => match resp.json::<InferResponseBody>().await {
+                Ok(val) => Ok(wasi_llm::InferencingResult {
+                    text: val.text,
+                    usage: wasi_llm::InferencingUsage {
+                        prompt_token_count: val.usage.prompt_token_count,
+                        generated_token_count: val.usage.generated_token_count,
+                    },
+                }),
+                Err(err) => Err(wasi_llm::Error::RuntimeError(format!(
+                    "Failed to deserialize response: {err}"
+                ))),
+            },
+            Err(err) => Err(wasi_llm::Error::RuntimeError(err.to_string())),
```

Comment:
We might want to provide a bit of additional context to this error since it might not be clear that this is from the call to `POST /infer`